pkg/apis/istio/v1alpha1/istioinstaller_types.proto

```diff
@@ -0,0 +1,585 @@
+field.
+
+3. k8s resource overlays
+
+Once a manifest is rendered from InstallSpec, a further specialization can be applied by specifying k8s resource
+overlays. The concept is similar to kustomize, where JSON patches are applied for object paths. This allows
+customization at the lowest level and eliminates the need to create ad-hoc template parameters.
+
+EXAMPLES
+
+1. Default Istio install
+
+spec: ~
+
+2. Default remote profile install
+
+spec:
+  baseSpecPath: istio-remote.yaml
+
+3. Default install with telemetry disabled
+
+spec:
+  telemetry:
+    enabled: false
+
+4. Default install with each feature installed to different namespace and security components in separate namespaces
+
+spec:
+  traffic_management:
+    namespace: istio-traffic-management
+  policy:
+    namespace: istio-policy
+  telemetry:
+    namespace: istio-telemetry
+  config_management:
+    namespace: istio-config-management
+  security:
+    citadel:
+      namespace: istio-citadel
+    cert_manager:
+      namespace: istio-cert-manager
+    node_agent:
+      namespace: istio-node-agent
+
+5. Default install with specialized k8s settings for pilot
+
+spec:
+  traffic_management:
+    pilot:
+        k8s:
+          resources:
+            limits:
+              cpu: 444m
+              memory: 333Mi
+            requests:
+              cpu: 222m
+              memory: 111Mi
+          readinessProbe:
+            failureThreshold: 44
+            initialDelaySeconds: 11
+            periodSeconds: 22
+            successThreshold: 33
+
+6. Default install with values.yaml customizations for proxy
+
+spec:
+  traffic_management:
+    proxy:
+      valuesOverrides:
+      - global.proxy.enableCoreDump: true
+      - global.proxy.dnsRefreshRate: 10s
+
+7. Default install with modification to container flag in galley
+
+spec:
+  configuration_management:
+    galley:
+      k8s:
+        overlays:
+        - apiVersion: extensions/v1beta1
+          kind: Deployment
+          name: istio-galley
+          patches:
+          - path: spec.template.spec.containers.[name:galley].command.[--livenessProbeInterval]
+            value: --livenessProbeInterval=123s
+*/
+
+
+message IstioInstall {
+    InstallSpec spec = 3;
+    InstallStatus status = 4;
+}
+
+// InstallSpec defines the desired state of IstioInstaller.
+// The spec is a used to define a customization of the default values that are supplied with an Istio charts release.
+// It is grouped at the top level by feature, where behavior of Istio functional areas is specified.
+// Each feature contains components, where k8s resource level defaults can be overridden.
+// Because the spec is a customization API, specifying an empty InstallSpec results in a default Istio installation.
+message InstallSpec {
+    // Prefix added to all namespaces for any installed component.
+    string default_namespace_prefix = 5;
+
+    // Selection and configuration of core Istio features.
```

Comment:
Perhaps add: if you don't specify any feature, the default will be used.

Reply:
That applies transitively to every field under InstallSpec, because this is an overlay spec. I stated that in InstallSpec comment above, but maybe you have a suggestion how I can emphasize this/make it more clear?

Reply:
How about something like `Selection and configuration of core Istio features over defaults"

Reply:
Why wouldn't you use the namespace where the CR is installed as the default?  I believe this is the _natural_ way to implement the operator.

Reply:
I think you're missing the point, you install the CR and a control plane is installed around it.  The CR does not need to be collocated with the operator itself.  The operator should be capable of installing multiple control planes (and we currently support this in Maistra, for multitenancy, with some caveats).

To clarify the CR namespace not the operator namespace.  Think of it like installing a deployment, the repica set and pods are created in the same namespace as the deployment resource.  Same paradigm.

Reply:
How will the operator find its CR if it's not in the same namespace?

Reply:
It can be configured to watch any or all namespaces.  It just depends on what you specify for the watch.  Currently, [operator.yaml is configured to watch its own namespace](https://github.com/istio/operator/blob/master/deploy/operator.yaml#L24-L27), but this can be changed to an empty string, in which case it watches all, or set to `istio-system`, in which case it watches `istio-system`.  There are ways to dynamically configure this when using [OLM](https://github.com/operator-framework/operator-lifecycle-manager).

Reply:
You could do that but it seems unnecessary - to me it makes sense to have the operator CR and resources in the same namespace because administratively the operator and its config belong to the same domain, whereas the Istio control plane components are in different admin domains. See also my other point about moving the config out of the operator and into mesh configmap or component config crds, and only keeping the install footprint in this crd.

Reply:
Actually, there is a good reason to _not_have the operator and CR collocated.  The operator will be running with a service account that has elevated privileges, which would bleed into the control plane namespace if they were collocated.  This allows users with lower privileges to install and manage a control plane.

Reply:
What's the difference between the operator CR and the istio CRs?  My assumption is that the control plane would be installed beside the IstioControlPlane CR.  If there's a requirement to deploy components into separate namespaces, then this would support that.  I don't have a problem with that.

I think one needs to be careful about that though, as it could have an impact on the network configuration.  For example, in OpenShift one can configure the default networking such that NetworkPolicy is being used and individual namespaces are segregated by default.  Deploying components to different namespaces would then require joining the networks of those namespaces so they can talk to each other.  That said, it would be up to the operator to detect the underlying networking, adding correctly configured NetworkPolicy resources so the control plane can talk to itself.  In a multitenant environment, containing multiple control planes, work is also required to join the user/application namespaces into the mesh.  This is something we're working on downstream in [Maistra/istio-operator](https://github.com/Maistra/istio-operator).

Reply:
There are three groups of CRs we've mentioned here. I've put suggested namespaces for each in [brackets].
1. operator CR [istio-operator-config-namespace]
2. Istio control plane CRs [istio-control, istio-security etc.]. Actually there are no CRs here, these are all standard k8s R's. Grouping these by feature (and possibly component) into ns allows separation of admin domains for these. This is a pretty basic feature of the new installer @costinm is working on and there's demand from customers for this type of feature admin domain separation. If you look at those charts, they do include explicit ns config so that the control plane can talk to itself.
3. Istio config CRs [istio-config]. This is user config so likely to be in one ns. 
So that leaves operator ns and that could be either in istio-operator-config-namespace or a separate ns - I don't have an opinion, but it should not be istio-config or any other the Istio control plane namespaces.

Reply:
Okay, so what is the operator CR, what does it describe?

Reply:
Exactly what we are looking at in this PR, meaning the enablement/namespaces of features and components and k8s resources for them (and possibly a few other config knobs like enable mtls that have impact on the installation). All other config items are passed through to mesh config. Right now this still results in charts being rendered with these values but the plan going forward (@costin, @mandarjog) is to gradually remove these items from the helm charts and have Istio control plane read it directly from mesh configmap.  
Note that we would still be able to able to use the operator to create/customize this mesh configmap, it just doesn't fall under the operator API we're defining here, but under a separate API.

Reply:
So this config, along with this operator, will be creating these other CRs as part of the installation process?  After which, human operators (administrators) would modify these other CRs?

Reply:
That's the general idea.